File: src/ormai/tools/generic.py
"""
Generic database tools.

These are the core tools that provide structured access to the database.
"""

from typing import Any

from pydantic import BaseModel, Field

from ormai.adapters.base import OrmAdapter
from ormai.core.context import RunContext
from ormai.core.dsl import (
    AggregateRequest,
    AggregateResult,
    BulkUpdateRequest,
    BulkUpdateResult,
    CreateRequest,
    CreateResult,
    DeleteRequest,
    DeleteResult,
    FilterClause,
    GetRequest,
    GetResult,
    IncludeClause,
    OrderClause,
    QueryRequest,
    QueryResult,
    UpdateRequest,
    UpdateResult,
)
from ormai.core.types import ModelMetadata, SchemaMetadata
from ormai.policy.models import ModelPolicy, Policy
from ormai.tools.base import Tool

# === Describe Schema Tool ===


class DescribeSchemaInput(BaseModel):
    """Input for describe_schema tool."""

    model: str | None = Field(
        default=None,
        description="Optional model name to get schema for. If not provided, returns all allowed models.",
    )


class SchemaDescription(BaseModel):
    """Output of describe_schema tool."""

    models: dict[str, Any] = Field(default_factory=dict)


class DescribeSchemaTool(Tool[DescribeSchemaInput, SchemaDescription]):
    """
    Tool to describe the database schema.

    Returns information about allowed models, fields, and relations
    based on the current policy.
    """

    name = "db.describe_schema"
    description = "Get the database schema including allowed models, fields, and relations."
    input_schema = DescribeSchemaInput

    def __init__(self, schema: SchemaMetadata, policy: Policy) -> None:
        self.schema = schema
        self.policy = policy

    async def execute(
        self,
        input: DescribeSchemaInput,
        ctx: RunContext,  # noqa: ARG002
    ) -> SchemaDescription:
        """Execute the describe_schema operation."""
        result: dict[str, Any] = {}

        if input.model:
            # Describe a specific model
            model_meta = self.schema.get_model(input.model)
            model_policy = self.policy.get_model_policy(input.model)

            if model_meta and model_policy and model_policy.allowed:
                result[input.model] = self._describe_model(model_meta, model_policy)
        else:
            # Describe all allowed models
            for model_name in self.policy.list_allowed_models():
                model_meta = self.schema.get_model(model_name)
                model_policy = self.policy.get_model_policy(model_name)

                if model_meta and model_policy:
                    result[model_name] = self._describe_model(model_meta, model_policy)

        return SchemaDescription(models=result)

    def _describe_model(
        self, model_meta: ModelMetadata, model_policy: ModelPolicy
    ) -> dict[str, Any]:
        """Build description for a single model."""
        # Get allowed fields
        fields = {}
        for field_name, field_meta in model_meta.fields.items():
            if model_policy.is_field_allowed(field_name):
                fields[field_name] = {
                    "type": field_meta.field_type.value,
                    "nullable": field_meta.nullable,
                    "primary_key": field_meta.primary_key,
                }

        # Get allowed relations
        relations = {}
        for rel_name, rel_meta in model_meta.relations.items():
            rel_policy = model_policy.relations.get(rel_name)
            if rel_policy is None or rel_policy.allowed:
                relations[rel_name] = {
                    "target": rel_meta.target_model,
                    "type": rel_meta.relation_type,
                }

        return {
            "table": model_meta.table_name,
            "fields": fields,
            "relations": relations,
            "readable": model_policy.readable,
            "writable": model_policy.writable,
        }


# === Query Tool ===


class QueryInput(BaseModel):
    """Input for query tool."""

    model: str = Field(..., description="The model to query")
    select: list[str] | None = Field(
        default=None, description="Fields to select (defaults to all allowed)"
    )
    where: list[dict[str, Any]] | None = Field(
        default=None, description="Filter conditions"
    )
    order_by: list[dict[str, Any]] | None = Field(
        default=None, description="Sort order"
    )
    take: int = Field(default=25, ge=1, le=100, description="Max rows to return")
    cursor: str | None = Field(default=None, description="Pagination cursor")
    include: list[dict[str, Any]] | None = Field(
        default=None, description="Relations to include"
    )


class QueryTool(Tool[QueryInput, QueryResult]):
    """
    Tool to query database records.

    Supports filtering, ordering, pagination, and relation includes.
    All queries are scoped and policy-enforced.
    """

    name = "db.query"
    description = "Query database records with filtering, ordering, and pagination."
    input_schema = QueryInput

    def __init__(self, adapter: OrmAdapter, policy: Policy, schema: SchemaMetadata) -> None:
        self.adapter = adapter
        self.policy = policy
        self.schema = schema

    async def execute(
        self,
        input: QueryInput,
        ctx: RunContext,
    ) -> QueryResult:
        """Execute the query operation."""
        # Build the request
        request = QueryRequest(
            model=input.model,
            select=input.select,
            where=[FilterClause.model_validate(f) for f in input.where] if input.where else None,
            order_by=[OrderClause.model_validate(o) for o in input.order_by] if input.order_by else None,
            take=input.take,
            cursor=input.cursor,
            include=[IncludeClause.model_validate(i) for i in input.include] if input.include else None,
        )

        # Compile and execute
        compiled = self.adapter.compile_query(request, ctx, self.policy, self.schema)
        return await self.adapter.execute_query(compiled, ctx)


# === Get Tool ===


class GetInput(BaseModel):
    """Input for get tool."""

    model: str = Field(..., description="The model to get from")
    id: Any = Field(..., description="The primary key value")
    select: list[str] | None = Field(
        default=None, description="Fields to select"
    )
    include: list[dict[str, Any]] | None = Field(
        default=None, description="Relations to include"
    )


class GetTool(Tool[GetInput, GetResult]):
    """
    Tool to get a single record by primary key.
    """

    name = "db.get"
    description = "Get a single record by its primary key."
    input_schema = GetInput

    def __init__(self, adapter: OrmAdapter, policy: Policy, schema: SchemaMetadata) -> None:
        self.adapter = adapter
        self.policy = policy
        self.schema = schema

    async def execute(
        self,
        input: GetInput,
        ctx: RunContext,
    ) -> GetResult:
        """Execute the get operation."""
        request = GetRequest(
            model=input.model,
            id=input.id,
            select=input.select,
            include=[IncludeClause.model_validate(i) for i in input.include] if input.include else None,
        )

        compiled = self.adapter.compile_get(request, ctx, self.policy, self.schema)
        return await self.adapter.execute_get(compiled, ctx)


# === Aggregate Tool ===


class AggregateInput(BaseModel):
    """Input for aggregate tool."""

    model: str = Field(..., description="The model to aggregate")
    operation: str = Field(..., description="Aggregation: count, sum, avg, min, max")
    field: str | None = Field(
        default=None, description="Field to aggregate (required for sum/avg/min/max)"
    )
    where: list[dict[str, Any]] | None = Field(
        default=None, description="Filter conditions"
    )


class AggregateTool(Tool[AggregateInput, AggregateResult]):
    """
    Tool to perform aggregations on database records.
    """

    name = "db.aggregate"
    description = "Perform aggregations (count, sum, avg, min, max) on records."
    input_schema = AggregateInput

    def __init__(self, adapter: OrmAdapter, policy: Policy, schema: SchemaMetadata) -> None:
        self.adapter = adapter
        self.policy = policy
        self.schema = schema

    async def execute(
        self,
        input: AggregateInput,
        ctx: RunContext,
    ) -> AggregateResult:
        """Execute the aggregation operation."""
        request = AggregateRequest(
            model=input.model,
            operation=input.operation,
            field=input.field,
            where=[FilterClause.model_validate(f) for f in input.where] if input.where else None,
        )

        compiled = self.adapter.compile_aggregate(request, ctx, self.policy, self.schema)
        return await self.adapter.execute_aggregate(compiled, ctx)


# =============================================================================
# MUTATION TOOLS (Phase 2)
# =============================================================================


class CreateInput(BaseModel):
    """Input for create tool."""

    model: str = Field(..., description="The model to create a record in")
    data: dict[str, Any] = Field(..., description="Field values for the new record")
    reason: str | None = Field(
        default=None,
        description="Reason for the mutation (may be required by policy)",
    )
    return_fields: list[str] | None = Field(
        default=None,
        description="Fields to return after creation",
    )


class CreateTool(Tool[CreateInput, CreateResult]):
    """
    Tool to create a new database record.

    Requires write permissions and validates against write policies.
    """

    name = "db.create"
    description = "Create a new database record."
    input_schema = CreateInput

    def __init__(self, adapter: OrmAdapter, policy: Policy, schema: SchemaMetadata) -> None:
        self.adapter = adapter
        self.policy = policy
        self.schema = schema

    async def execute(
        self,
        input: CreateInput,
        ctx: RunContext,
    ) -> CreateResult:
        """Execute the create operation."""
        # Build the request
        request = CreateRequest(
            model=input.model,
            data=input.data,
            reason=input.reason,
            return_fields=input.return_fields,
        )

        # Compile and execute
        compiled = self.adapter.compile_create(request, ctx, self.policy, self.schema)
        return await self.adapter.execute_create(compiled, ctx)


class UpdateInput(BaseModel):
    """Input for update tool."""

    model: str = Field(..., description="The model to update")
    id: Any = Field(..., description="The primary key of the record to update")
    data: dict[str, Any] = Field(..., description="Fields to update")
    reason: str | None = Field(
        default=None,
        description="Reason for the mutation",
    )
    return_fields: list[str] | None = Field(
        default=None,
        description="Fields to return after update",
    )


class UpdateTool(Tool[UpdateInput, UpdateResult]):
    """
    Tool to update a database record by primary key.

    Requires write permissions and validates against write policies.
    """

    name = "db.update"
    description = "Update a database record by its primary key."
    input_schema = UpdateInput

    def __init__(self, adapter: OrmAdapter, policy: Policy, schema: SchemaMetadata) -> None:
        self.adapter = adapter
        self.policy = policy
        self.schema = schema

    async def execute(
        self,
        input: UpdateInput,
        ctx: RunContext,
    ) -> UpdateResult:
        """Execute the update operation."""
        request = UpdateRequest(
            model=input.model,
            id=input.id,
            data=input.data,
            reason=input.reason,
            return_fields=input.return_fields,
        )

        compiled = self.adapter.compile_update(request, ctx, self.policy, self.schema)
        return await self.adapter.execute_update(compiled, ctx)


class DeleteInput(BaseModel):
    """Input for delete tool."""

    model: str = Field(..., description="The model to delete from")
    id: Any = Field(..., description="The primary key of the record to delete")
    reason: str | None = Field(
        default=None,
        description="Reason for the deletion",
    )
    hard: bool = Field(
        default=False,
        description="If True, perform hard delete instead of soft delete",
    )


class DeleteTool(Tool[DeleteInput, DeleteResult]):
    """
    Tool to delete a database record by primary key.

    By default, performs a soft delete if the model supports it.
    Requires write/delete permissions.
    """

    name = "db.delete"
    description = "Delete a database record by its primary key."
    input_schema = DeleteInput

    def __init__(self, adapter: OrmAdapter, policy: Policy, schema: SchemaMetadata) -> None:
        self.adapter = adapter
        self.policy = policy
        self.schema = schema

    async def execute(
        self,
        input: DeleteInput,
        ctx: RunContext,
    ) -> DeleteResult:
        """Execute the delete operation."""
        request = DeleteRequest(
            model=input.model,
            id=input.id,
            reason=input.reason,
            hard=input.hard,
        )

        compiled = self.adapter.compile_delete(request, ctx, self.policy, self.schema)
        return await self.adapter.execute_delete(compiled, ctx)


class BulkUpdateInput(BaseModel):
    """Input for bulk update tool."""

    model: str = Field(..., description="The model to update")
    ids: list[Any] = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Primary keys of records to update",
    )
    data: dict[str, Any] = Field(..., description="Fields to update on all records")
    reason: str | None = Field(
        default=None,
        description="Reason for the mutation",
    )


class BulkUpdateTool(Tool[BulkUpdateInput, BulkUpdateResult]):
    """
    Tool to update multiple database records by their primary keys.

    Safer than filter-based bulk updates because it requires explicit IDs.
    Requires write permissions and bulk operation permissions.
    """

    name = "db.bulk_update"
    description = "Update multiple database records by their primary keys."
    input_schema = BulkUpdateInput

    def __init__(self, adapter: OrmAdapter, policy: Policy, schema: SchemaMetadata) -> None:
        self.adapter = adapter
        self.policy = policy
        self.schema = schema

    async def execute(
        self,
        input: BulkUpdateInput,
        ctx: RunContext,
    ) -> BulkUpdateResult:
        """Execute the bulk update operation."""
        request = BulkUpdateRequest(
            model=input.model,
            ids=input.ids,
            data=input.data,
            reason=input.reason,
        )

        compiled = self.adapter.compile_bulk_update(request, ctx, self.policy, self.schema)
        return await self.adapter.execute_bulk_update(compiled, ctx)

File: ormai-ts/src/tools/generic.ts
/**
 * Generic database tools.
 *
 * These are the core tools that provide structured access to the database.
 */

import { z } from 'zod';
import type { RunContext } from '../core/context.js';
import {
  type AggregateResult,
  type BulkUpdateResult,
  type CreateResult,
  type DeleteResult,
  type GetResult,
  type QueryResult,
  type UpdateResult,
  AggregateRequestSchema,
  BulkUpdateRequestSchema,
  CreateRequestSchema,
  DeleteRequestSchema,
  FilterClauseSchema,
  GetRequestSchema,
  IncludeClauseSchema,
  OrderClauseSchema,
  QueryRequestSchema,
  UpdateRequestSchema,
} from '../core/dsl.js';
import type { SchemaMetadata } from '../core/types.js';
import type { OrmAdapter } from '../adapters/base.js';
import type { Policy } from '../policy/models.js';
import { PolicyUtils, ModelPolicyUtils } from '../policy/models.js';
import { BaseTool } from './base.js';

// =============================================================================
// TOOL LIMITS
// =============================================================================

/**
 * Tool limits for configuring constraints.
 */
export interface ToolLimits {
  /** Maximum number of rows returned by query (default: 100) */
  maxQueryRows: number;
  /** Maximum number of IDs in bulk update (default: 100) */
  maxBulkUpdateIds: number;
  /** Maximum number of IDs in bulk delete (default: 100) */
  maxBulkDeleteIds: number;
}

/**
 * Default tool limits.
 */
export const DEFAULT_TOOL_LIMITS: ToolLimits = {
  maxQueryRows: 100,
  maxBulkUpdateIds: 100,
  maxBulkDeleteIds: 100,
};

// =============================================================================
// DESCRIBE SCHEMA TOOL
// =============================================================================

const DescribeSchemaInputSchema = z.object({
  model: z
    .string()
    .optional()
    .describe('Optional model name to get schema for. If not provided, returns all allowed models.'),
});

type DescribeSchemaInput = z.infer<typeof DescribeSchemaInputSchema>;

interface SchemaDescription {
  models: Record<string, ModelDescription>;
}

interface ModelDescription {
  table: string;
  fields: Record<string, FieldDescription>;
  relations: Record<string, RelationDescription>;
  readable: boolean;
  writable: boolean;
}

interface FieldDescription {
  type: string;
  nullable: boolean;
  primaryKey: boolean;
}

interface RelationDescription {
  target: string;
  type: string;
}

/**
 * Tool to describe the database schema.
 */
export class DescribeSchemaTool extends BaseTool<DescribeSchemaInput, SchemaDescription> {
  readonly name = 'db.describe_schema';
  readonly description = 'Get the database schema including allowed models, fields, and relations.';
  readonly inputSchema = DescribeSchemaInputSchema;

  constructor(
    private readonly schema: SchemaMetadata,
    private readonly policy: Policy
  ) {
    super();
  }

  async execute(input: DescribeSchemaInput, _ctx: RunContext): Promise<SchemaDescription> {
    const result: Record<string, ModelDescription> = {};

    if (input.model) {
      const modelMeta = this.schema.models[input.model];
      const modelPolicy = PolicyUtils.getModelPolicy(this.policy, input.model);

      if (modelMeta && modelPolicy?.allowed) {
        result[input.model] = this.describeModel(input.model, modelMeta, modelPolicy);
      }
    } else {
      for (const modelName of PolicyUtils.listAllowedModels(this.policy)) {
        const modelMeta = this.schema.models[modelName];
        const modelPolicy = PolicyUtils.getModelPolicy(this.policy, modelName);

        if (modelMeta && modelPolicy) {
          result[modelName] = this.describeModel(modelName, modelMeta, modelPolicy);
        }
      }
    }

    return { models: result };
  }

  private describeModel(
    _modelName: string,
    modelMeta: SchemaMetadata['models'][string],
    modelPolicy: NonNullable<ReturnType<typeof PolicyUtils.getModelPolicy>>
  ): ModelDescription {
    const fields: Record<string, FieldDescription> = {};
    for (const [fieldName, fieldMeta] of Object.entries(modelMeta.fields)) {
      if (ModelPolicyUtils.isFieldAllowed(modelPolicy, fieldName)) {
        fields[fieldName] = {
          type: fieldMeta.fieldType,
          nullable: fieldMeta.nullable,
          primaryKey: fieldMeta.primaryKey,
        };
      }
    }

    const relations: Record<string, RelationDescription> = {};
    for (const [relName, relMeta] of Object.entries(modelMeta.relations)) {
      const relPolicy = modelPolicy.relations[relName];
      if (!relPolicy || relPolicy.allowed) {
        relations[relName] = {
          target: relMeta.targetModel,
          type: relMeta.relationType,
        };
      }
    }

    return {
      table: modelMeta.tableName,
      fields,
      relations,
      readable: modelPolicy.readable,
      writable: modelPolicy.writable,
    };
  }
}

// =============================================================================
// QUERY TOOL
// =============================================================================

const QueryInputSchema = z.object({
  model: z.string().describe('The model to query'),
  select: z.array(z.string()).optional().describe('Fields to select (defaults to all allowed)'),
  where: z.array(z.record(z.unknown())).optional().describe('Filter conditions'),
  orderBy: z.array(z.record(z.unknown())).optional().describe('Sort order'),
  take: z.number().int().min(1).default(25).describe('Max rows to return'),
  cursor: z.string().optional().describe('Pagination cursor'),
  include: z.array(z.record(z.unknown())).optional().describe('Relations to include'),
});

type QueryInput = z.infer<typeof QueryInputSchema>;

/**
 * Tool to query database records.
 */
export class QueryTool extends BaseTool<QueryInput, QueryResult> {
  readonly name = 'db.query';
  readonly description = 'Query database records with filtering, ordering, and pagination.';
  readonly inputSchema = QueryInputSchema;

  constructor(
    private readonly adapter: OrmAdapter,
    private readonly policy: Policy,
    private readonly schema: SchemaMetadata,
    private readonly limits: ToolLimits = DEFAULT_TOOL_LIMITS
  ) {
    super();
  }

  async execute(input: QueryInput, ctx: RunContext): Promise<QueryResult> {
    // Validate take against limits
    if (input.take > this.limits.maxQueryRows) {
      throw new Error(`take value ${input.take} exceeds maximum of ${this.limits.maxQueryRows}`);
    }

    const request = QueryRequestSchema.parse({
      model: input.model,
      select: input.select,
      where: input.where?.map((f) => FilterClauseSchema.parse(f)),
      orderBy: input.orderBy?.map((o) => OrderClauseSchema.parse(o)),
      take: input.take,
      cursor: input.cursor,
      include: input.include?.map((i) => IncludeClauseSchema.parse(i)),
    });

    const compiled = this.adapter.compileQuery(request, ctx, this.policy, this.schema);
    return this.adapter.executeQuery(compiled, ctx);
  }
}

// =============================================================================
// GET TOOL
// =============================================================================

const GetInputSchema = z.object({
  model: z.string().describe('The model to get from'),
  id: z.unknown().describe('The primary key value'),
  select: z.array(z.string()).optional().describe('Fields to select'),
  include: z.array(z.record(z.unknown())).optional().describe('Relations to include'),
});

type GetInput = z.infer<typeof GetInputSchema>;

/**
 * Tool to get a single record by ID.
 */
export class GetTool extends BaseTool<GetInput, GetResult> {
  readonly name = 'db.get';
  readonly description = 'Get a single record by its primary key.';
  readonly inputSchema = GetInputSchema;

  constructor(
    private readonly adapter: OrmAdapter,
    private readonly policy: Policy,
    private readonly schema: SchemaMetadata
  ) {
    super();
  }

  async execute(input: GetInput, ctx: RunContext): Promise<GetResult> {
    const request = GetRequestSchema.parse({
      model: input.model,
      id: input.id,
      select: input.select,
      include: input.include?.map((i) => IncludeClauseSchema.parse(i)),
    });

    const compiled = this.adapter.compileGet(request, ctx, this.policy, this.schema);
    return this.adapter.executeGet(compiled, ctx);
  }
}

// =============================================================================
// AGGREGATE TOOL
// =============================================================================

const AggregateInputSchema = z.object({
  model: z.string().describe('The model to aggregate'),
  operation: z.enum(['count', 'sum', 'avg', 'min', 'max']).describe('Aggregation operation'),
  field: z.string().optional().describe('Field to aggregate (required for sum/avg/min/max)'),
  where: z.array(z.record(z.unknown())).optional().describe('Filter conditions'),
});

type AggregateInput = z.infer<typeof AggregateInputSchema>;

/**
 * Tool to perform aggregations.
 */
export class AggregateTool extends BaseTool<AggregateInput, AggregateResult> {
  readonly name = 'db.aggregate';
  readonly description = 'Perform aggregation operations (count, sum, avg, min, max).';
  readonly inputSchema = AggregateInputSchema;

  constructor(
    private readonly adapter: OrmAdapter,
    private readonly policy: Policy,
    private readonly schema: SchemaMetadata
  ) {
    super();
  }

  async execute(input: AggregateInput, ctx: RunContext): Promise<AggregateResult> {
    const request = AggregateRequestSchema.parse({
      model: input.model,
      operation: input.operation,
      field: input.field,
      where: input.where?.map((f) => FilterClauseSchema.parse(f)),
    });

    const compiled = this.adapter.compileAggregate(request, ctx, this.policy, this.schema);
    return this.adapter.executeAggregate(compiled, ctx);
  }
}

// =============================================================================
// CREATE TOOL
// =============================================================================

const CreateInputSchema = z.object({
  model: z.string().describe('The model to create'),
  data: z.record(z.unknown()).describe('Field values for the new record'),
  reason: z.string().optional().describe('Reason for the mutation'),
  returnFields: z.array(z.string()).optional().describe('Fields to return after creation'),
});

type CreateInput = z.infer<typeof CreateInputSchema>;

/**
 * Tool to create a new record.
 */
export class CreateTool extends BaseTool<CreateInput, CreateResult> {
  readonly name = 'db.create';
  readonly description = 'Create a new database record.';
  readonly inputSchema = CreateInputSchema;

  constructor(
    private readonly adapter: OrmAdapter,
    private readonly policy: Policy,
    private readonly schema: SchemaMetadata
  ) {
    super();
  }

  async execute(input: CreateInput, ctx: RunContext): Promise<CreateResult> {
    const request = CreateRequestSchema.parse({
      model: input.model,
      data: input.data,
      reason: input.reason,
      returnFields: input.returnFields,
    });

    const compiled = this.adapter.compileCreate(request, ctx, this.policy, this.schema);
    return this.adapter.executeCreate(compiled, ctx);
  }
}

// =============================================================================
// UPDATE TOOL
// =============================================================================

const UpdateInputSchema = z.object({
  model: z.string().describe('The model to update'),
  id: z.unknown().describe('The primary key of the record to update'),
  data: z.record(z.unknown()).describe('Fields to update'),
  reason: z.string().optional().describe('Reason for the mutation'),
  returnFields: z.array(z.string()).optional().describe('Fields to return after update'),
});

type UpdateInput = z.infer<typeof UpdateInputSchema>;

/**
 * Tool to update a record.
 */
export class UpdateTool extends BaseTool<UpdateInput, UpdateResult> {
  readonly name = 'db.update';
  readonly description = 'Update a database record by its primary key.';
  readonly inputSchema = UpdateInputSchema;

  constructor(
    private readonly adapter: OrmAdapter,
    private readonly policy: Policy,
    private readonly schema: SchemaMetadata
  ) {
    super();
  }

  async execute(input: UpdateInput, ctx: RunContext): Promise<UpdateResult> {
    const request = UpdateRequestSchema.parse({
      model: input.model,
      id: input.id,
      data: input.data,
      reason: input.reason,
      returnFields: input.returnFields,
    });

    const compiled = this.adapter.compileUpdate(request, ctx, this.policy, this.schema);
    return this.adapter.executeUpdate(compiled, ctx);
  }
}

// =============================================================================
// DELETE TOOL
// =============================================================================

const DeleteInputSchema = z.object({
  model: z.string().describe('The model to delete from'),
  id: z.unknown().describe('The primary key of the record to delete'),
  reason: z.string().optional().describe('Reason for the deletion'),
  hard: z.boolean().default(false).describe('If true, perform hard delete'),
});

type DeleteInput = z.infer<typeof DeleteInputSchema>;

/**
 * Tool to delete a record.
 */
export class DeleteTool extends BaseTool<DeleteInput, DeleteResult> {
  readonly name = 'db.delete';
  readonly description = 'Delete a database record (soft delete by default).';
  readonly inputSchema = DeleteInputSchema;

  constructor(
    private readonly adapter: OrmAdapter,
    private readonly policy: Policy,
    private readonly schema: SchemaMetadata
  ) {
    super();
  }

  async execute(input: DeleteInput, ctx: RunContext): Promise<DeleteResult> {
    const request = DeleteRequestSchema.parse({
      model: input.model,
      id: input.id,
      reason: input.reason,
      hard: input.hard,
    });

    const compiled = this.adapter.compileDelete(request, ctx, this.policy, this.schema);
    return this.adapter.executeDelete(compiled, ctx);
  }
}

// =============================================================================
// BULK UPDATE TOOL
// =============================================================================

const BulkUpdateInputSchema = z.object({
  model: z.string().describe('The model to update'),
  ids: z.array(z.unknown()).min(1).describe('Primary keys of records to update'),
  data: z.record(z.unknown()).describe('Fields to update on all records'),
  reason: z.string().optional().describe('Reason for the mutation'),
});

type BulkUpdateInput = z.infer<typeof BulkUpdateInputSchema>;

/**
 * Tool to bulk update records.
 */
export class BulkUpdateTool extends BaseTool<BulkUpdateInput, BulkUpdateResult> {
  readonly name = 'db.bulk_update';
  readonly description = 'Update multiple records by their primary keys.';
  readonly inputSchema = BulkUpdateInputSchema;

  constructor(
    private readonly adapter: OrmAdapter,
    private readonly policy: Policy,
    private readonly schema: SchemaMetadata,
    private readonly limits: ToolLimits = DEFAULT_TOOL_LIMITS
  ) {
    super();
  }

  async execute(input: BulkUpdateInput, ctx: RunContext): Promise<BulkUpdateResult> {
    // Validate ids count against limits
    if (input.ids.length > this.limits.maxBulkUpdateIds) {
      throw new Error(`Bulk update IDs count ${input.ids.length} exceeds maximum of ${this.limits.maxBulkUpdateIds}`);
    }

    const request = BulkUpdateRequestSchema.parse({
      model: input.model,
      ids: input.ids,
      data: input.data,
      reason: input.reason,
    });

    const compiled = this.adapter.compileBulkUpdate(request, ctx, this.policy, this.schema);
    return this.adapter.executeBulkUpdate(compiled, ctx);
  }
}

// =============================================================================
// FACTORY FUNCTION
// =============================================================================

/**
 * Options for creating generic tools.
 */
export interface GenericToolsOptions {
  adapter: OrmAdapter;
  policy: Policy;
  schema: SchemaMetadata;
  includeWrite?: boolean;
  limits?: Partial<ToolLimits>;
}

/**
 * Create all generic database tools.
 */
export function createGenericTools(options: GenericToolsOptions): BaseTool<unknown, unknown>[] {
  const { adapter, policy, schema, includeWrite = false, limits } = options;
  const effectiveLimits = { ...DEFAULT_TOOL_LIMITS, ...limits };

  const tools: BaseTool<unknown, unknown>[] = [
    new DescribeSchemaTool(schema, policy),
    new QueryTool(adapter, policy, schema, effectiveLimits),
    new GetTool(adapter, policy, schema),
    new AggregateTool(adapter, policy, schema),
  ];

  if (includeWrite || policy.writesEnabled) {
    tools.push(
      new CreateTool(adapter, policy, schema),
      new UpdateTool(adapter, policy, schema),
      new DeleteTool(adapter, policy, schema),
      new BulkUpdateTool(adapter, policy, schema, effectiveLimits)
    );
  }

  return tools;
}
